### PR TITLE
Refactors how authentication providers are managed.

### PR DESCRIPTION
**Pull request by kevgliss**
This PR accomplished two main goals:

1) Make it easier to add new ways of adding and manging authentication providers
2) Decouples our frontend code from the PKCE auth flow

Together, both of these changes should make it much easier to enable and support different auth flow, including disabling auth altogether. See #95 

We also introduce an authentication plugin system via the `AuthenticationProviderPlugin` base class. This is a simple interface that additional backend authentication flows can leverage. From a user's perspective swapping out authentication providers should be as easy as modifying the new config variable `DISPATCH_AUTHENTICATION_PROVIDER_SLUG` (we keep pkce as the default, setting this to "" disables all auth).

cc @willbengtson

### CODE REVIEW

**Comment by diogomatsubara** [comment]
I have a branch based on this deployed at https://master-7rqtwti-2hounbfgfsnu6.eu-3.platformsh.site/dashboard with authentication disabled if people want to try it out and give feedback.

**Comment by kevgliss** [comment]
@diogomatsubara that's awesome! We were thinking about hosting it somewhere for folks to play with but hadn't got a chance. I'm also planning to generate some fake data so the demo is a bit more interesting.

**Comment by diogomatsubara** [comment]
@kevgliss if you land some fake data I'd be happy to keep that instance updated for others to play with. I been trying to put some data in manually but that's get tedious really fast. :-)

File: src/dispatch/config.py [hunk 1]
```diff
@@ -44,19 +44,24 @@ def __str__(self) -> str:
...
 # authentication
-DISPATCH_PKCE_AUTH_ENABLED = config("DISPATCH_PKCE_AUTH", cast=bool, default=True)
-DISPATCH_HEADER_AUTH_ENABLED = config("DISPATCH_HEADER_AUTH", cast=bool, default=False)
-
-DISPATCH_AUTH_HEADER_KEY = config("DISPATCH_AUTH_HEADER", default=None)
-JWKS_URL = config("JWKS_URL", default=None)
+DISPATCH_AUTHENTICATION_PROVIDER_SLUG = config(
+    "DISPATCH_AUTHENTICATION_PROVIDER_SLUG", default="dispatch-auth-provider-pkce"
+)
+VUE_APP_DISPATCH_AUTHENTICATION_PROVIDER_SLUG = DISPATCH_AUTHENTICATION_PROVIDER_SLUG
```
Comment (mvilanova) [inline src/dispatch/config.py]: Should we get rid of `VUE_APP_DISPATCH_AUTHENTICATION_PROVIDER_SLUG` if we're always setting it to `DISPATCH_AUTHENTICATION_PROVIDER_SLUG`, and use `DISPATCH_AUTHENTICATION_PROVIDER_SLUG` instead?

Comment (kevgliss) [reply src/dispatch/config.py]: Unfortunately, Vue will only pick up envvars prefixed with `VUE_APP` to be injected into the vue app. I think this is the best of both worlds (only having to set one var).

**Pull request merged**